symengine/set_funcs.cpp

```diff
@@ -0,0 +1,164 @@
+#ifndef SYMENGINE_SET_FUNCS_H
+#define SYMENGINE_SET_FUNCS_H
+
+#include <symengine/visitor.h>
+#include <symengine/functions.h>
+
+namespace SymEngine
+{
+
+class SupVisitor : public BaseVisitor<SupVisitor>
+{
+private:
+    RCP<const Basic> sup_;
+
+public:
+    SupVisitor() {}
+
+    void bvisit(const Basic &x){};
+
+    void bvisit(const Set &x)
+    {
+        throw SymEngineException(
+            "Set not partially ordered: supremum undefined");
+    };
+
+    void bvisit(const Reals &x)
+    {
+        sup_ = infty(1);
+    };
+
+    void bvisit(const Rationals &x)
+    {
+        sup_ = infty(1);
+    };
+
+    void bvisit(const Integers &x)
+    {
+        sup_ = infty(1);
+    };
+
+    void bvisit(const Interval &x)
+    {
+        sup_ = x.get_end();
+    };
+
+    void bvisit(const FiniteSet &x)
+    {
+        const set_basic &container = x.get_container();
+        vec_basic v(container.begin(), container.end());
+        sup_ = max(v);
+    };
+
+    void bvisit(const Union &x)
+    {
+        RCP<const Basic> m = infty(-1);
+        for (auto &a : x.get_container()) {
+            a->accept(*this);
+            m = max({m, sup_});
```

Comment:
Shouldn't this be calling `sup` instead of `max`?

Reply:
No, I don't think so. A `Union` contains multiple sets `a`. We visit the next set in line 57 to get its supremum in `sup_`. The supremum up until the current iteration will be the max of this and the previous max in `m`. We could build a vector of these values and after the loop take the max. Perhaps it would be faster.

Reply:
You are correct. It should be `max`. Can you build a vector?

Reply:
ok